### PR TITLE
[Inference]Fused the gate and up proj in mlp，and optimized the autograd process.

### PR DESCRIPTION
**Pull request by yuehuayingxueluo**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
Part of optimization described in https://github.com/hpcaitech/ColossalAI/issues/5323

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
pytest：
image
benchmark:
|--------|---------|-----|--------------------------|
| 256 | 128 | 16 | 1021.90 -> 1146.61 |
| 256 | 128 | 32 | 1747.18 -> 1950.64 |
| 256 | 128 | 64 | 2294.75 -> 2873.97 |
| 1024 | 128 | 16 | 891.69 -> 999.88 |
| 1024 | 128 | 32 | 1207.50 -> 1479.34 |
| 1024 | 128 | 64 | 1381.41 -> 1805.08 |

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/core/engine.py [hunk 1]
```diff
@@ -113,9 +113,6 @@ def _shardformer(
             model_policy (Policy): The policy to shardformer model which is determined by the model type.
             stage_manager (PipelineStageManager, optional): Used to manage pipeline stages. Defaults to None.
             tp_group (ProcessGroupMesh, optional): Used to manage the process TP group mesh. Defaults to None.
-
-        Returns:
-            nn.Module: _description_
```
Comment (CjhHa1) [inline colossalai/inference/core/engine.py:118]: maybe use a new description rather than delete

Comment (yuehuayingxueluo) [reply colossalai/inference/core/engine.py:118]: fixed

File: colossalai/inference/core/engine.py [hunk 2]
```diff
@@ -149,25 +150,25 @@ def generate(
         Returns:
             List[str]: Inference result returned by one generation.
         """
+        with torch.no_grad():
```
Comment (FrankLeeeee) [inline colossalai/inference/core/engine.py:153]: `torch.inference_mode` is favored over `torch.no_grad()` now.

File: colossalai/kernel/triton/flash_decoding.py [hunk 3]
```diff
@@ -261,6 +261,9 @@ def flash_decoding_attention(
     # NOTE use `triton.next_power_of_2` here to utilize the cache mechanism of triton
     # To optimize, revise batching/scheduling to batch 2^n sequences in a batch (preferred)
     grid = (triton.next_power_of_2(bsz), num_heads, triton.cdiv(triton.next_power_of_2(max_seq_len_in_batch), BLOCK_KV))
+    output = torch.empty((bsz, num_heads, head_dim), dtype=q.dtype, device=q.device) if output is None else output
+    grid_1 = (triton.next_power_of_2(bsz), num_heads)
```
Comment (FrankLeeeee) [inline colossalai/kernel/triton/flash_decoding.py:265]: Why call it grid_1?

File: colossalai/kernel/triton/rms_layernorm.py [hunk 4]
```diff
@@ -66,4 +65,4 @@ def rms_layernorm(x, weight, eps, norm_output=None):
 
         # enqueue kernel
         _rmsnorm_kernel[(M,)](x, y, weight, x.stride(0), N, eps, BLOCK_SIZE=BLOCK_SIZE, num_warps=num_warps)
-        return y
+        return y, x
```
Comment (FrankLeeeee) [inline colossalai/kernel/triton/rms_layernorm.py:68]: Why do you need to return the input as input is not modified at all.

Comment (yuehuayingxueluo) [reply colossalai/kernel/triton/rms_layernorm.py:68]: In this way，we can avoid implicit DToD memcopy operation.

Comment (FrankLeeeee) [reply colossalai/kernel/triton/rms_layernorm.py:68]: What is the difference between these two cases?

- case 1:
```python
out = rms_layernorm(x, weight)

# do something on x
```

- case 2
```

```python
out, x = rms_layernorm(x, weight)

# do something on x
```
```

File: colossalai/kernel/triton/flash_decoding.py (hunk 3 above)
Comment (FrankLeeeee) [reply colossalai/kernel/triton/flash_decoding.py:265]: ya, but I still do not know why the variable is named as `grid_1`, `grid` and `grid_1` have trivial difference in terms of their naming.

Comment (yuehuayingxueluo) [reply colossalai/kernel/triton/flash_decoding.py:265]: My previous idea was to calculate the required grid for each kernel before executing the Triton kernel.

File: colossalai/kernel/triton/rms_layernorm.py (hunk 4 above)
Comment (FrankLeeeee) [reply colossalai/kernel/triton/rms_layernorm.py:68]: I suspect that this is what actually happens.

File: colossalai/kernel/triton/flash_decoding.py (hunk 3 above)
Comment (FrankLeeeee) [reply colossalai/kernel/triton/flash_decoding.py:265]: What I want to discuss is why we name it with `_1`, what meaning does `_1` carry?

File: colossalai/kernel/triton/rms_layernorm.py (hunk 4 above)
Comment (FrankLeeeee) [reply colossalai/kernel/triton/rms_layernorm.py:68]: As `rms_layernorm` does not touch on `x` at all, the input argument should remain unchanged no matter if `x` is returned.

File: colossalai/kernel/triton/flash_decoding.py (hunk 3 above)
Comment (yuehuayingxueluo) [reply colossalai/kernel/triton/flash_decoding.py]: My previous intention was to distinguish between the names of the two grids, but the naming convention I used might not have been rigorous enough.

Comment (FrankLeeeee) [reply colossalai/kernel/triton/flash_decoding.py]: Ok, do avoid casual naming for the sake of clarity.

File: colossalai/kernel/triton/rms_layernorm.py (hunk 4 above)
Comment (yuehuayingxueluo) [reply colossalai/kernel/triton/rms_layernorm.py]: fixed

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**